core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,214 @@
+  }
+
+  /**
+   * Get the full set of nodes that are blacklisted.  Unlike other methods in this class, this *IS*
+   * thread-safe -- no lock required on a taskScheduler.
+   */
+  def nodeBlacklist(): Set[String] = {
+    _nodeBlacklist.get()
+  }
+
+  def isNodeBlacklisted(node: String): Boolean = {
+    nodeIdToBlacklistExpiryTime.contains(node)
+  }
+
+  def removeExecutor(executorId: String): Unit = {
+    // we intentionally do not clean up executors that are already blacklisted, so that if another
+    // executor on the same node gets blacklisted, we can blacklist the entire node.
+    executorIdToFailureCount -= executorId
+  }
+}
+
+
+private[scheduler] object BlacklistTracker extends Logging {
+
+  /**
+   * Return true if the blacklist is enabled, based on the following order of preferences:
+   * 1. Is it specifically enabled or disabled?
+   * 2. Is it enabled via the legacy timeout conf?
+   * 3. Use the default for the spark-master:
+   *   - off for local mode
+   *   - on for distributed modes (including local-cluster)
+   */
+  def isBlacklistEnabled(conf: SparkConf): Boolean = {
+    conf.get(config.BLACKLIST_ENABLED) match {
+      case Some(isEnabled) =>
+        isEnabled
+      case None =>
+        // if they've got a non-zero setting for the legacy conf, always enable the blacklist,
+        // otherwise, use the default based on the cluster-mode (off for local-mode, on otherwise).
+        val legacyKey = config.BLACKLIST_LEGACY_TIMEOUT_CONF.key
+        conf.get(config.BLACKLIST_LEGACY_TIMEOUT_CONF) match {
+          case Some(legacyTimeout) if legacyTimeout == 0 =>
+            logWarning(s"Turning off blacklisting due to legacy configuaration:" +
+              s" $legacyKey == 0")
+            false
+          case Some(legacyTimeout) =>
+              // mostly this is necessary just for tests, since real users that want the blacklist
+              // will get it anyway by default
+              logWarning(s"Turning on blacklisting due to legacy configuration:" +
+                s" $legacyKey > 0")
+              true
+          case None =>
+            // local-cluster is *not* considered local for these purposes, we still want the
+            // blacklist enabled by default
+            !Utils.isLocalMaster(conf)
+        }
+    }
+  }
+
+  def getBlacklistExpiryTime(conf: SparkConf): Long = {
+    val timeoutConf = conf.get(config.BLACKLIST_EXPIRY_TIMEOUT_CONF)
+    val legacyTimeoutConf = conf.get(config.BLACKLIST_LEGACY_TIMEOUT_CONF)
+    (timeoutConf, legacyTimeoutConf) match {
+      case (Some(x), _) => x
+      case (None, Some(y)) => y
+      case (None, None) =>
+        Utils.timeStringAsMs("1h")
+    }
+  }
+}
+
+/** Failures for one executor, within one taskset */
+private[scheduler] final class FailureStatus(val node: String) {
+  /** index of the tasks in the taskset that have failed on this executor. */
+  val tasksWithFailures = HashSet[Int]()
```

Comment:
its organized this way so we can quickly sum up all the failures on one executor, to see if it puts the executors over the threshold.  That could be expensive if it wasn't organized by executor -- on every failure, you'd need to go through all the task failures.  I ran some performance test where there was some slowdown without this, though it was pretty minor.  I think it was just when you had a real error in user code on a large cluster -- lots of task failures pile, on lots of different executors, which leads to some non-trivial work in each task start / end event.

Perhaps just some renaming here would do the trick?  FailureStatus is pretty generic.  Especially the way its used in TaskSetManager.  I'll try to clean it up a bit and see what you think ... or maybe you could point out the places that are still confusing.